File: src/main/java/szewek/mcflux/blocks/BlockEnergyMachine.java
package szewek.mcflux.blocks;

import net.minecraft.block.properties.PropertyEnum;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.IStringSerializable;
import net.minecraft.util.NonNullList;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import szewek.fl.FL;
import szewek.mcflux.tileentities.TileEntityEnergyMachine;

public final class BlockEnergyMachine extends BlockMCFluxContainer {
	private static final AxisAlignedBB DEF_AABB = new AxisAlignedBB(0.25, 0.25, 0.25, 0.75, 0.75, 0.75);
	public static final PropertyEnum<Variant> VARIANT = PropertyEnum.create("variant", Variant.class);

	public BlockEnergyMachine() {
		super();
		setHardness(0.5F);
	}

	@Override
	public void getDrops(NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune) {
		drops.add(new ItemStack(this, 1, state.getValue(VARIANT).ordinal()));
	}

	@Override
	public int damageDropped(IBlockState state) {
		return state.getValue(VARIANT).ordinal();
	}

	@Override
	public TileEntity createNewTileEntity(World w, int m) {
		TileEntityEnergyMachine teem = new TileEntityEnergyMachine();
		teem.setModuleId(m);
		return teem;
	}

	@Override
	public void getSubBlocks(CreativeTabs tab, NonNullList<ItemStack> list) {
		for (int i = 0; i < Variant.ALL_VARIANTS.length; i++)
			list.add(new ItemStack(this, 1, i));
	}

	@Override
	public IBlockState getStateFromMeta(int meta) {
		return this.getDefaultState().withProperty(VARIANT, Variant.ALL_VARIANTS[meta % Variant.ALL_VARIANTS.length]);
	}

	@Override
	public int getMetaFromState(IBlockState state) {
		return state.getValue(VARIANT).ordinal();
	}

	@Override
	protected BlockStateContainer createBlockState() {
		return new BlockStateContainer(this, VARIANT);
	}

	@Override
	public AxisAlignedBB getBoundingBox(IBlockState state, IBlockAccess source, BlockPos pos) {
		return DEF_AABB;
	}

	@Override
	public boolean isFullCube(IBlockState state) {
		return false;
	}

	@Override
	public boolean isOpaqueCube(IBlockState state) {
		return false;
	}

	@Override
	public boolean onBlockActivated(World w, BlockPos bp, IBlockState ibs, EntityPlayer p, EnumHand h, EnumFacing f, float x, float y, float z) {
		boolean b = h == EnumHand.MAIN_HAND && FL.isItemEmpty(p.getHeldItem(h));
		if (b && !w.isRemote) {
			TileEntity te = w.getTileEntity(bp);
			if (te != null && te instanceof TileEntityEnergyMachine)
				((TileEntityEnergyMachine) te).switchSideTransfer(f);
		}
		return b;
	}

	public enum Variant implements IStringSerializable {
		ENERGY_DIST("energy_dist"), CHUNK_CHARGER("chunk_charger");

		public static final Variant[] ALL_VARIANTS;
		public final String name;

		Variant(String n) {
			name = n;
		}

		@Override
		public String getName() {
			return name;
		}

		public static String nameFromStack(ItemStack is) {
			return ALL_VARIANTS[is.getMetadata() % ALL_VARIANTS.length].name;
		}

		static {
			ALL_VARIANTS = Variant.values();
		}
	}
}

File: src/main/java/szewek/mcflux/blocks/BlockEnergyMachine.kt
package szewek.mcflux.blocks

import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.IStringSerializable
import net.minecraft.util.NonNullList
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import szewek.fl.block.BlockContainerModeled
import szewek.mcflux.tileentities.TileEntityEnergyMachine

class BlockEnergyMachine : BlockContainerModeled() {
	init {
		setHardness(0.5f)
	}

	override fun getDrops(drops: NonNullList<ItemStack>, world: IBlockAccess?, pos: BlockPos?, state: IBlockState, fortune: Int) {
		drops += ItemStack(this, 1, state.getValue(VARIANT).ordinal)
	}

	override fun damageDropped(state: IBlockState): Int = state.getValue(VARIANT).ordinal

	override fun createNewTileEntity(w: World, m: Int): TileEntity? {
		val teem = TileEntityEnergyMachine()
		teem.moduleId = m
		return teem
	}

	override fun getSubBlocks(tab: CreativeTabs, list: NonNullList<ItemStack>) {
		for (i in Variant.ALL_VARIANTS.indices)
			list += ItemStack(this, 1, i)
	}

	override fun getStateFromMeta(meta: Int) =
			defaultState.withProperty(VARIANT, Variant.ALL_VARIANTS[meta % Variant.ALL_VARIANTS.size])

	override fun getMetaFromState(state: IBlockState) = state.getValue(VARIANT).ordinal
	override fun createBlockState() = BlockStateContainer(this, VARIANT)
	override fun getBoundingBox(state: IBlockState, source: IBlockAccess, pos: BlockPos) = DEF_AABB
	override fun isFullCube(state: IBlockState) = false
	override fun isOpaqueCube(state: IBlockState) = false

	override fun onBlockActivated(w: World?, bp: BlockPos?, ibs: IBlockState?, p: EntityPlayer?, h: EnumHand?, f: EnumFacing?, x: Float, y: Float, z: Float): Boolean {
		val b = h == EnumHand.MAIN_HAND && p!!.getHeldItem(h).isEmpty
		if (b && !w!!.isRemote) {
			val te = w.getTileEntity(bp!!)
			if (te != null && te is TileEntityEnergyMachine)
				te.switchSideTransfer(f!!)
		}
		return b
	}

	enum class Variant constructor(private val vname: String) : IStringSerializable {
		ENERGY_DIST("energy_dist"), CHUNK_CHARGER("chunk_charger");

		override fun getName() = vname

		companion object {

			@JvmField
			val ALL_VARIANTS: Array<Variant> = Variant.values()

			fun nameFromStack(stk: ItemStack): String {
				return ALL_VARIANTS[stk.metadata % ALL_VARIANTS.size].vname
			}

		}
	}

	companion object {
		private val DEF_AABB = AxisAlignedBB(0.25, 0.25, 0.25, 0.75, 0.75, 0.75)
		val VARIANT = PropertyEnum.create("variant", Variant::class.java)
	}
}
